### ISSUE TITLE
PHP Warning: mb_strlen(): Unknown encoding "" in /var/www/.../vendor/respect/validation/library/Rules/Length.php:57

### ISSUE THREAD

**Issue by wouterds** [Respect/Validation#681]
Sometimes my error reporter is sending me these errors.

**Comment by PascalZajac** [comment]
Just wanted to add to this report, we've seen a lot of these warnings in the last month. Our particular validation setup is:

```php
        $this->add([
            'id' => v::intVal(),
            'email' =>  v::email()->notEmpty(),
            'first_name' => v::stringType()->length(0, 255, true),
            'last_name' => v::stringType()->length(0, 255, true),
            'source' => v::stringType()->length(0, 255, true),
            'order_id' => v::oneOf([v::intVal(), v::nullType()])
        ]);
```

(The `add` method just defines an array of validators that are iterated over at runtime)

The problem seems to happen when the `source` field is `null`, but I'm not sure what sort of request causes `mb_detect_encoding` to fail.

**Comment by GusAntoniassi** [comment]
I am also having problems with this warning, but in my case it's related to the ISO 8859-1 encoding.

With the following file (must be saved as ISO 8859-1 for the error to be thrown -- the actual values are coming from a database with the same encoding):
```php
use Respect\Validation\Validator as v;

var_dump(v::length(1, 50)->validate('Foo D´Bar'));
```
I get "Warning: mb_strlen(): Unknown encoding &quot;&quot; in 
/var/www/html/vendor/respect/validation/library/Rules/Length.php on line 57" and the validation returns as `false`.

**Comment by AnrDaemon** [comment]
This happening because you are using `mb_detect_encoding`.
Do not, never ever use it. Forget it exists.
Change it to 
```php
return mb_strlen($input, ini_get("mbstring.internal_encoding") ?: ini_get("internal_encoding") ?: ini_get("default_charset") ?: "ISO-8859-1");
```

**Comment by AnrDaemon** [comment]
Even more correct solution would have been passing encoding to the validator explicitly.

**Comment by henriquemoody** [comment]
I'm closing issues that are too old, and I'm not really looking at them all. If this is still relevant, please comment here and I will reopen this issue.

Thanks for reporting! 🐼

**Comment by AnrDaemon** [comment]
This is still an issue.
```less
$ ag --php --search-binary --pager=less -i 'mb_detect'
Rules/Charset.php
20:use function mb_detect_encoding;
57:        $detectedEncoding = mb_detect_encoding($input, $this->charset, true);

Rules/Contains.php
19:use function mb_detect_encoding;
77:        $encoding = (string) mb_detect_encoding($haystack);

Rules/EndsWith.php
18:use function mb_detect_encoding;
73:        $encoding = (string) mb_detect_encoding($input);
88:        $encoding = (string) mb_detect_encoding($input);

Rules/In.php
18:use function mb_detect_encoding;
79:        return mb_stripos($this->haystack, $inputString, 0, (string) mb_detect_encoding($input>
97:        return mb_strpos($this->haystack, $inputString, 0, (string) mb_detect_encoding($inputS>

Rules/Length.php
25:use function mb_detect_encoding;
92:            return (int) mb_strlen($input, (string) mb_detect_encoding($input));

Rules/Lowercase.php
17:use function mb_detect_encoding;
39:        return $input === mb_strtolower($input, (string) mb_detect_encoding($input));

Rules/StartsWith.php
17:use function mb_detect_encoding;
71:        return mb_stripos($input, $this->startValue, 0, (string) mb_detect_encoding($input)) =>
83:        return mb_strpos($input, $this->startValue, 0, (string) mb_detect_encoding($input)) ==>

Rules/Uppercase.php
17:use function mb_detect_encoding;
39:        return $input === mb_strtoupper($input, (string) mb_detect_encoding($input));
```

**Comment by AnrDaemon** [comment]
(To be clear, the very use of `mb_detect_encoding` is an issue. This function simply not meant to be used for single-byte or UNICODE encodings.)

**Comment by henriquemoody** [comment]
This fix is already available on version 2.2.1.

Thank you for reporting! And thank you @AnrDaemon for taking a step further on explaining better the issue. 🐼
